compiler/rustc_hir_analysis/src/delegation.rs

```diff
@@ -76,127 +77,300 @@ fn fn_kind<'tcx>(tcx: TyCtxt<'tcx>, def_id: DefId) -> FnKind {
     }
 }
 
+type OwnParams = Vec<ty::GenericParamDef>;
+
+struct GenericsBuilder<'tcx> {
+    tcx: TyCtxt<'tcx>,
+    sig_id: DefId,
+    parent: Option<DefId>,
+    own_params: OwnParams,
+}
+
+impl<'tcx> GenericsBuilder<'tcx> {
+    fn new(tcx: TyCtxt<'tcx>, sig_id: DefId) -> GenericsBuilder<'tcx> {
+        GenericsBuilder { tcx, sig_id, parent: None, own_params: vec![] }
+    }
+
+    fn with_parent(mut self, parent: DefId) -> Self {
+        self.parent = Some(parent);
+        self
+    }
+
+    fn with_own_generics(mut self, def_id: DefId) -> Self {
+        let generics = self.tcx.generics_of(def_id);
+        self.own_params.append(&mut generics.own_params.clone());
+        self
+    }
+
+    fn with_generics(mut self, def_id: DefId) -> Self {
+        let generics = self.tcx.generics_of(def_id);
+        if let Some(parent_def_id) = generics.parent {
+            self = self.with_own_generics(parent_def_id);
+        }
+        self.with_own_generics(def_id)
+    }
+
+    fn skip_self(mut self) -> Self {
+        self.own_params = self.own_params.iter().skip(1).cloned().collect_vec();
+        self
+    }
+
+    fn build(mut self) -> ty::Generics {
+        // Lifetime parameters must be declared before type and const parameters.
+        // Therefore, When delegating from a free function to a associated function,
+        // generic parameters need to be reordered:
+        //
+        // trait Trait<'a, A> {
+        //     fn foo<'b, B>(...) {...}
+        // }
+        //
+        // reuse Trait::foo;
+        // desugaring:
+        // fn foo<'a, 'b, This: Trait<'a, A>, A, B>(...) {
+        //     Trait::foo(...)
+        // }
+        self.own_params.sort_by_key(|key| key.kind.is_ty_or_const());
+
+        let param_def_id_to_index =
+            self.own_params.iter().map(|param| (param.def_id, param.index)).collect();
+
+        let (parent_count, has_self) = if let Some(def_id) = self.parent {
+            let parent_generics = self.tcx.generics_of(def_id);
```

Comment:
I'm not sure which logic is put into `fn with_(own_)generics` and which into `fn build` and why.
In both places we request some `generics_of` from `def_id`, `sig_id` and the parent and do something with them.

Maybe it would be better `fn with_(own_)generics` were "inert" and only set some flags, and `fn build` would perform all the work?

Reply:
Same with `PredicatesBuilder`.

Reply:
All logic was moved into `build` and is now toggled by `InheritanceKind`.